File: String/commonPrefix.py
def longestCommonPrefix(strs):
    prefix = ""

    if len(strs) == 0:
        return ""

    if len(strs) == 1:
        return strs[0]

    # Sort the list to place a word with the smallest size to the first place
    def lenComparator(i):
        return len(i)

    strs.sort(key=lenComparator)

    for i in range(0, len(strs[0])):
        iChar = strs[0][i]

        shouldContinue = True

        for j in range(1, len(strs)):
            if strs[j][i] != iChar:
                shouldContinue = False
                break

        if shouldContinue == False:
            return prefix
        else:
            prefix += iChar

    return prefix


# print(longestCommonPrefix([]))
print(longestCommonPrefix(["flower", "flow", "flight"]))


#
# Runtime: O(N lonN) + O(N^2) => O(N * lonN + N ^ 2) => O(N^2)
#
#

File: String/commonPrefix.js
/**
 * @param {string[]} strs
 * @return {string}
 */

const longestCommonPrefix = strs => {
  let prefix = "";

  if (strs.length === 0) return "";
  if (strs.length === 1) return strs[0];

  strs = strs.sort((a, b) => a.length - b.length); // O(N logN)

  for (i = 0; i < strs[0].length; i++) {
    const char = strs[0][i];

    let shouldContinue = true;

    for (j = 1; j < strs.length; j++) {
      if (strs[j][i] !== char) {
        shouldContinue = false;
        break;
      }
    }

    if (!shouldContinue) {
      return prefix;
    } else {
      prefix += char;
    }
  }

  return prefix;
};

console.log(longestCommonPrefix(["flower", "flow", "flight"]));

/*
 *
 * Runtime complexity: (N logN) + (N * N) -> (N logN + N^2) => O(N^2);
 *
 *
 * Space complexity: N logN
 */
